### ISSUE TITLE
Auto-tagging tickets to be part of the release

### ISSUE THREAD

**Issue by timwright12** [department-of-veterans-affairs/va-mobile-app#6807]
Currently we have to manually tag an issue with the release it will be a part of after a PR is merged into develop. This is a pretty we established process, so we should be able to automate it.

**The process should work like this:**

1. Issue is created
2. Work starts
3. PR opens, ticket is connected to issue
4. PR approved
5. PR Merged into develop
6. Issue connected is updated with the release milestone

**Comment by timwright12** [comment]
@BeccaTupaj I think this would be a good one to get in the pipeline with more teams coming onboard (cc @jbergman-oddball )

**Comment by jbergman-oddball** [comment]
I think we need to decide on what mechanism we want to use here.  Tags or Milestones.

They both have their pros and cons

# Pros

## Tags
- Can be applied at any time to the issue. 
- Easily searchable and filterable.
- Probably a lot more

## Milestones
- Provides a milestone view.  [Example](https://github.com/department-of-veterans-affairs/va-mobile-app/milestone/11?closed=1)
- More of a project view of the issues. 
![Image](https://github.com/user-attachments/assets/56cbd957-ab16-4fa3-bf6e-ee18209df7fb)
- Probably also a lot more



# Cons

## Tags
- The biggest one is if we "tag" the issues with the release tag it may kick off the automation. The git workflows  look for this. [Example](https://github.com/department-of-veterans-affairs/va-mobile-app/blob/develop/.github/workflows/release_build.yml#L9)
- This tag is already created as the automation flows use it.  We may introduce issues if we tag again.  (above)


## Milestones
- None come to mind.  It is isolated from tags.  Will have its own separate view.



If we decide tags, we need to be aware of the risks.  If we choose milestones, we would not want to auto add these at the end of the sprint. We would want to add this during planning or refinement so we are able to track their progress. 

cc: @BeccaTupaj @timwright12

**Comment by BeccaTupaj** [comment]
I agree with using Milestones

**Comment by jbergman-oddball** [comment]
I may have found a way to do this.  We can create a workflow that takes issues (tied to a sprint or a label) and apply a milestone that way. I will look into mocking this up.  I will need a tag or another way to be able top pull the correct issues.

**Comment by timwright12** [comment]
After my short discovery I did find ways to:
- Trigger a workflow with a PR is merged/closed ([ref](https://stackoverflow.com/questions/60710209/trigger-github-actions-only-when-pr-is-merged))
- Flag only code that touches the app ([ref](https://github.com/department-of-veterans-affairs/va-mobile-app/blob/develop/.github/workflows/documentation_test_build.yml#L8))
- Get the issue number attached to a PR ([ref](https://github.com/nearform-actions/github-action-check-linked-issues))
- Add that issue to a milestone (gh issue edit 5 --milestone v1.1.0) ([ref](https://cli.github.com/manual/gh_issue_edit))

**Open issues:**

- the format returned from github-action-check-linked-issues isn't just the number, so that needs to be cleaned
- need a way to get the next up milestone

The logic for this action should be:

1. Pull request is closed/merged
2. If the code updated was inside VAMobile/src, run the workflow otherwise skip
3. If no linked issue, exit
4. If linked issue, get the issue number
5. Locate the appropriate milestone (maybe check the due date?)
6. Use the github CLI to add the issue to the milestone (`gh issue edit 5 --milestone v1.1.0`)

This should give us a list of issues included in each release.

After this portion is released we can start a new ticket to address issues that live outside our repo (experience teams). Going to unblock this now.

**Comment by BeccaTupaj** [comment]
Thanks @timwright12 this is great!

**Comment by BeccaTupaj** [comment]
@timwright12 and @jbergman-oddball What is the next step needed for this ticket?

**Comment by jbergman-oddball** [comment]
I found a way to get the next milestone
```
gh api /repos/department-of-veterans-affairs/va-mobile-app/milestones | jq '.[].number' | sort -u | head -n 1
```
This will pull open milestones and get the current open list sort them and only return the first result.

We can use a variation of this command to pull the title which will return our `vx.x.x`
```
 gh api /repos/department-of-veterans-affairs/va-mobile-app/milestones | jq '.[].title' | sort -u | head -n 1 | sed 's/^"//; s/"$//'
```
gh cli does not have the ability to do this (from what I can see with my version) so the API requests are necessary.

From here we can run a PATCH to close the milestone at the end of the release.
```
gh api -X PATCH /repos/department-of-veterans-affairs/va-mobile-app/milestones/<number> -f state=closed
```

Will be testing this theory in my test repo.

**Comment by jbergman-oddball** [comment]
To get the issue data from the connected issues list, we can use these commands

This will return the repo and issue number.  We can use this to only get the # and issue number
```
echo "department-of-veterans-affairs/va-mobile-app#6807" | sed 's/.*#/\#/'
#6807
```

If we don't want to `#`, we would use this
```
echo "department-of-veterans-affairs/va-mobile-app#6807" | sed 's/.*#//'
6807
```

**Comment by jbergman-oddball** [comment]
After we retrieve the issue number and the milestone, we can run:
```
gh issue edit <issue> --milestone <milestone>
```
Then when we close the issue we would run:

`gh api -X PATCH /repos/department-of-veterans-affairs/va-mobile-app/milestones/<number> -f state=closed`


@BeccaTupaj @timwright12

**Comment by timwright12** [comment]
> This will pull open milestones and get the current open list sort them and only return the first result.
Do we know how "first" is defined here? Creation date? Due date? Currently this is the same for us, but might not always be if something was created out of order

> From here we can run a PATCH to close the milestone at the end of the release.
@BeccaTupaj can you confirm this is when we want the milestone closed?

> Then when we close the issue we would run:
We shouldn't need extra automation to close an issue, it should close automatically when the PR is merged (current behavior)

**Comment by jbergman-oddball** [comment]
> Do we know how "first" is defined here? Creation date? Due date? Currently this is the same for us, but might not always be if something was created out of order
---
My thought on this was sorting them so the list will always be in order.  That way if it is created out of order, the sort in the above command will bring it inline.

> We shouldn't need extra automation to close an issue, it should close automatically when the PR is merged (current behavior)
---
Copy. I will omit this.

**Comment by BeccaTupaj** [comment]
I believe the milestone should be closed with the release is complete (i.e. released to production)

**Comment by jbergman-oddball** [comment]
I was able to get `act` working to test the GH action before deployment. This is extremely helpful with this particular case due to the fact that I need a PR created, approved and merged when I test this.  with `act` I can send a mock PR payload and update the fields accordingly to test the different scenarios. GH actions are completing successfully and I am able to run `gh api` to pull the data from the repo to get the milestones.
[Act Reference](https://github.com/nektos/act)

**Comment by jbergman-oddball** [comment]
This is working.
From test repo:
Non Linked Issue
[pre merge](https://github.com/jbergman-oddball/va-test/actions/runs/15074938946/job/42380084849)
[post merge](https://github.com/jbergman-oddball/va-test/actions/runs/15074950525/job/42380120195)
[relevant pr](https://github.com/jbergman-oddball/va-test/pull/10)

Linked issue
[relevant pr](https://github.com/jbergman-oddball/va-test/pull/11)

@BeccaTupaj @timwright12
